src/tt_scheduler.erl

```diff
@@ -0,0 +1,148 @@
+%%%-------------------------------------------------------------------
+%%% @author Wojciech Pachuta
+%%%-------------------------------------------------------------------
+-module(tt_scheduler).
+-author("Wojciech Pachuta").
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0, schedule/3, cancel_schedule/1]).
+
+%% gen_server callbacks
+-export([init/1,
+  handle_call/3,
+  handle_cast/2,
+  handle_info/2,
+  terminate/2,
+  code_change/3]).
+
+%% auxiliary functions
+-export([time_keeper_function/4,  prompt_publishing/1]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {refs :: #{reference() => pid()}}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec(start_link() ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+  gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+-spec(schedule(calendar:datetime(), calendar:datetime(), calendar:time()) -> atom()).
+schedule(StartTime, EndTime, PublishInterval) ->
+  gen_server:call(?SERVER, {schedule, StartTime, EndTime, PublishInterval}).
+
+-spec(cancel_schedule(reference()) -> no_return()).
+cancel_schedule(Ref) ->
+  gen_server:cast(?SERVER, {cancel_schedule, Ref}).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec(init(Args :: term()) ->
+  {ok, State :: #state{}} | {ok, State :: #state{}, timeout() | hibernate} |
+  {stop, Reason :: term()} | ignore).
+init([]) ->
+  {ok, #state{refs = maps:new()}}.
+
+-spec(handle_call(Request :: term(), From :: {pid(), Tag :: term()},
```

Comment:
It is not a good idea to provide an 'original' specifications of the callback functions. It is better to write one that matches your specific callback implementation, like 

```erlang
-spec handle_call({schedule, StartTime :: time(), EndTime :: time(), Interval :: time()) -> {reply, reference(), state()}.
```